### PR TITLE
Add an exercise on UBSan

### LINKED ISSUE #243: UBSan exercise
Use UBSan to run broken examples.

### PR DESCRIPTION
**Pull request by hageboeck**
Add a bug-ridden program that runs mostly without observable issues, but is full of undefined behaviour.

Using UBSan, all of these issues get revealed at run time.

Fix #243

### CODE REVIEW

**Review by bernhardmgruber — APPROVED** [review]
I gave it a quick look and the example seems fine!

Great addition, thank you!

File: exercises/ubsan/solution/undefinedBehaviour.cpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+ * Recompile with "-fsanitize=undefined", and observe that almost every second line contains a serious bug.
+ * Try to understand what's wrong.
+ */
+
+int runTests() {
+  int arr[] = {1, 2, 3, 4};
+  std::cout << "arr[4]=" << arr[4] << "\n";             // Array overrun
+
+  unsigned short s = 1;
+  std::cout << "s << 33=" << (s << 33) << "\n";         // We cannot shift a type with 32 bits by 33 bits.
```
Comment (sponce) [inline exercises/ubsan/solution/undefinedBehaviour.cpp:43]: And even less a 16 bits one !

**Review by sponce — APPROVED** [review]

File: exercises/ubsan/solution/undefinedBehaviour.cpp (hunk 1 above)
Comment (hageboeck) [reply exercises/ubsan/solution/undefinedBehaviour.cpp:43]: Interestingly (@sponce), it didn't trigger when I shifted by 17 bits. So I read up on cppreference, and it turns out the integral promotion is applied before shifting. The right-hand side is `int`, so it was correct ... 🙂

Comment (sponce) [reply exercises/ubsan/solution/undefinedBehaviour.cpp:43]: indeed, that's interesting. But then I would use an unsigned int as I do not want to have to explain these details here.

Comment (hageboeck) [reply exercises/ubsan/solution/undefinedBehaviour.cpp:43]: OK, good suggestion. I merged in the mean, time, but I will quickly edit the file in-place now.

**Pull request merged**